docs/src/best-practices-js.md

```diff
@@ -0,0 +1,171 @@
+
+
+```js
+👎 page.locator('button.buttonIcon.episode-actions-later')
+```
+
+Use locators that are resilient to changes in the DOM.
+
+```js
+👍 page.getByRole('button', { name: 'submit' })
+```
+## Generate locaters
+
+Playwright has a [test generator](./codegen.md) that can generate tests and pick locators for you. It will look at your page and figure out the best locator, prioritizing role, text and test id locators. If the generator finds multiple elements matching the locator, it will improve the locator to make it resilient and uniquely identify the target element, so you don't have to worry about failing tests due to locators.
+
+#### Use the codegen to generate locators
+
+To pick a locator you can run the codegen command and click on the pick locator button. Then hover over your page in the browser window and click on the element you want to pick. You can then copy and paste this locator into your code. You can also use the codegen to record a test for you.
+
+```bash
+npx playwright codegen
+```
+#### Use the VS Code extension to generate locators
+
+You can also use the [VS Code Extension](./getting-started-vscode.md) to generate locators as well as record a test. The VS Code extension also gives you a great developer experience when writing, running and debugging tests.
+
+## Use web first assertions
+
+Assertions are a way to verify that the expected result and the actual result matched or not. By using [web first assertions](./test-assertions.md) Playwright will wait until the expected condition is met. For Example when testing a toast message, if you click a button that makes a toast message appear you can test the toast message is there. If the toast takes half a second to appear Playwright will wait. Web assertions such as `toBeVisible()` will wait and retry.
+
+```js
+👍 await expect(page.getByText('welcome')).toBeVisible();
+```
+
+#### Don't use manual assertions
+
+Don't use manual assertions that are not awaiting the expect. In the code below the await is inside the expect rather than before it. When using assertions such as `isVisible()` the test wont wait a single second, it will just check the locator is there and return immediately. Use web first assertions such as `toBeVisible()` instead.
+
+```js
+👎 expect(await page.getByText('welcome').isVisible()).toBe(true);
+```
+
+Use web first assertions such as `toBeVisible()` instead.
+
+```js
+👍 await expect(page.getByText('welcome')).toBeVisible();
+```
+## Configure post-mortem debugging
+
+Use the playwright [trace viewer](./trace-viewer.md) for local debugging and post mortem debugging instead of videos and screenshots. The trace viewer gives you a full trace of your tests as a local PWA that can easily be shared. With the trace viewer you can view the timeline, inspect DOM snapshots for each action, view network requests and more.
```

Comment:
```suggestion
Debug your tests live in VSCode. <add a link/text here>

For CI failures, use the playwright [trace viewer](./trace-viewer.md) instead of videos and screenshots. The trace viewer gives you a full trace of your tests as a local PWA that can easily be shared. With the trace viewer you can view the timeline, inspect DOM snapshots for each action, view network requests and more.
```